### PR TITLE
8343417: (fs) BasicFileAttributeView.setTimes uses microsecond precision with NOFOLLOW_LINKS

### PR DESCRIPTION
**Pull request by bplb**
Add support for setting the last access and last modification times of symbolic links with nanosecond precision on Linux where the system call `utimensat(2)` is available.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343417](https://bugs.openjdk.org/browse/JDK-8343417): (fs) BasicFileAttributeView.setTimes uses microsecond precision with NOFOLLOW_LINKS (**Bug** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21886/head:pull/21886` \
`$ git checkout pull/21886`

Update a local copy of the PR: \
`$ git checkout pull/21886` \
`$ git pull https://git.openjdk.org/jdk.git pull/21886/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21886`

View PR using the GUI difftool: \
`$ git pr show -t 21886`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21886.diff">https://git.openjdk.org/jdk/pull/21886.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21886#issuecomment-2455542080)
</details>

### CODE REVIEW

File: src/java.base/unix/classes/sun/nio/fs/UnixNativeDispatcher.java [hunk 1]
```diff
@@ -391,6 +391,22 @@ static void lutimes(UnixPath path, long times0, long times1)
     private static native void lutimes0(long pathAddress, long times0, long times1)
         throws UnixException;
 
+    /**
+     * utimensat(int fd, const char* path,
+     *           const struct timeval times[2], int flags)
+     *
+     * We hard code fd to FD_ATCWD and flags to AT_SYMLINK_NOFOLLOW.
+     */
+    static void utimensat(UnixPath path, long times0, long times1)
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixNativeDispatcher.java:400]: The *at methods that are supposed to map 1-1 to the similarly named sys calls so I'm a bit uncomfortable having this one work differently and hardcode dirfd  to FD_ATCWD. I wonder if we could generate a value for FD_ATCWD in UnixConstants.java.template so that the use site would more closely match the sys call.

Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixNativeDispatcher.java:400]: > I wonder if we could generate a value for FD_ATCWD in UnixConstants.java.template so that the use site would more closely match the sys call.

Sure, no problem.

Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixNativeDispatcher.java]: So changed in 26feef6.

File: src/java.base/unix/classes/sun/nio/fs/UnixFileAttributeViews.java [hunk 2]
```diff
@@ -76,13 +76,16 @@ public void setTimes(FileTime lastModifiedTime,
...
-                    useLutimes = lutimesSupported() &&
-                        UnixFileAttributes.get(file, false).isSymbolicLink();
+                if (!followLinks &&
+                    UnixFileAttributes.get(file, false).isSymbolicLink()) {
+                    useUtimensat = utimensatSupported();
+                    if (!useUtimensat)
+                        useLutimes = lutimesSupported();
                 }
-                if (!useLutimes) {
+                if (!useUtimensat && !useLutimes) {
```
[the remark this answers is not in the document]
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixFileAttributeViews.java:88]: Indeed UnixFileAttributes.get calls lstat. Calling symlink.openForAttributeAccess(false) however yields
```
java.nio.file.FileSystemException: symlink: No such file or directory or unable to access attributes of symbolic link
```
According to [symlink(7)](https://www.man7.org/linux/man-pages/man7/symlink.7.html):
```
The last access and last modification timestamps of a symbolic link can be changed using utimensat(2) or lutimes(3).
```
It looks like futimens will not work with a symlink's fd so that utimensat is necessary here to obtain nsec precision. I don't see a way to close the window between lstat and utimensat, which anyway was already there for the lutimes case.

[the remark this answers is not in the document]
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixFileAttributeViews.java:88]: I don't think that the logic in 08f5b7d is quite correct yet. It might be possible if followLinks is false and the file is not a symlink, that lutimes would be used even though futimens is supported, thus decreasing the timestamp resolution.

**Comment by bplb** [comment]
> I think what you have is correct now but it's hard to follow as it's one of 4 sys calls. [...] I realise this is adding to the work on this change but right now it's hard for anyone to touch this method due to the many cases.

It definitely needs reworking but I do not know the details as yet.

**Review by AlanBateman — APPROVED** [review]
Discussed with Brian about some refactoring setTimes to make it more maintainable, and maybe drop the use of methods that only support micro seconds precision The AIX port may hold us back, not sure. So approving for now, there will be cleanup to follow.
